### ISSUE TITLE
Adding root_path breaks admin page

### ISSUE THREAD

**Issue by mmzeynalli** [aminalaee/sqladmin#538]
### Checklist

- [X] The bug is reproducible against the latest release or `master`.
- [X] There are no similar issues or pull requests to fix it yet.

### Describe the bug

I have multiple endpoints and admin panel. I want my endpoints to be in root path of `/api/v1/`, but admin to be just `/admin/`. So, I added `app_configs["root_path"] = f"/api/v{settings.APP_VERSION}"`, and admin panel breaks down.

### Steps to reproduce the bug

```
app_configs: dict[str, Any] = {"title": "Itinerago API"}
app_configs["root_path"] = f"/api/v{settings.APP_VERSION}"

app = FastAPI(**app_configs)
app.include_router(auth_router, prefix="/auth", tags=["Auth"])
app.include_router(core_router, prefix="", tags=["Main Logic"])


admin = Admin(app, engine)

# admin.add_view-s here
```

Commenting out second line fixes admin, however, it does not give desired output.

### Expected behavior

APIs will have `/api/v1/` prefix, but, admin will not. Admin will load without a problem.

### Actual behavior

Admin breaks down completely:

![image](https://github.com/aminalaee/sqladmin/assets/33568903/054cb6bb-15b4-4b66-9136-5921d94b5eaf)

and

![image](https://github.com/aminalaee/sqladmin/assets/33568903/d578ca0a-ba09-4436-89b8-5fabd66bc670)

Links are also broken:
In the first image, when you click one of the models it goes to: /api/v1/admin/model/list

![image](https://github.com/aminalaee/sqladmin/assets/33568903/0b78544b-46fe-4841-9694-bd41a3473bb8)

### Environment

Windows, Python 3.10, Dockerized Fastapi
sqladmin version 0.12.0

**Comment by aminalaee** [comment]
Hey,
not sure if I get your question right.
I can think of two points:
- Use can use `Admin(base_url=...)` to configure the URL.
- In your app you don't se the root_path in path, instead that's the prefix in your routers.

Generally I think this makes sense since you want to handle multiple API versionings in your app, running at the same time so setting the version on the whole app might not be a good idea.

**Comment by mmzeynalli** [comment]
Default base_url suits me, however, whenever set any root_path, admin gets messy like I showed in screenshots. You second point would work, but, it is hassle to add `api/v1/` to every app's router. I think admin should be independent of root_path and have its url.

**Comment by aminalaee** [comment]
Anyway if you want to do that you can just modify admin.app which is just another FastAPI app so you can do any config you want.

**Comment by mmzeynalli** [comment]
How? Just adding root_path breaks admin panel. Is it expected?

**Comment by aminalaee** [comment]
I haven't tested this before, doesn't setting this work?

```
admin.app["root_path"] = ...
```

**Comment by mmzeynalli** [comment]
Just checked, no it does not work. Change the line to:
`admin.app.root_path = ""`, but as `app` passed to admin is the global fastapi app, I am overriding global root_path => All endpoints are without /api/v1/

**Comment by VladBein** [comment]
@mmzeynalli have you solved this problem?

**Comment by mmzeynalli** [comment]
To be honest no, unfortunately, have been super busy for the last couple of months.

**Comment by tanguyMichardiere** [comment]
A workaround I found is to use a middleware:

```python
@app.middleware("http")
async def fix_admin_root_path(request, call_next):
    if request.url.path.startswith("/admin/"):
        request.scope["path"] = app.root_path + request.url.path
    return await call_next(request)
```

**Comment by logicli0n** [comment]
<img width="1680" alt="Screenshot 2024-07-12 at 12 45 45 AM" src="https://github.com/aminalaee/sqladmin/assets/50915489/84d538f8-c679-48e2-87ad-c1375173587f">

I set root_path="/api/v1"
Admin page works in /api/v1/admin and normal.
But when i use ajax_lookup in admin, ajax_lookup made request http://localhost:8070/admin/user/ajax/lookup?name=role&term=a without adding root_path and it give 404 Not Found

**Comment by logicli0n** [comment]
The problem is in here https://github.com/aminalaee/sqladmin/blob/main/sqladmin/application.py#L226? 
Can u add root_path before base_url? @aminalaee

**Comment by aminalaee** [comment]
@logicli0n Sure. Are you setting root path with admin.app.root_path ?

**Comment by logicli0n** [comment]
@aminalaee  I'm setting root path on my main app(in Fastapi).

**Comment by logicli0n** [comment]
Did u fix @aminalaee?

**Comment by vahidzhe** [comment]
I have the same problem.

**Comment by MaximDementyev** [comment]
The problem listed in the description is solved by a simple code:

```
app = FastAPI() # Additional root application without endpoints
main_app = FastAPI() # app with endpoints

admin = Admin(app, engine)
app.mount('/api/v1', main_app)

main_app.include_router(auth_router, prefix="/auth", tags=["Auth"])
```
@mmzeynalli how appropriate is it to rewrite the logic of pathfinding (mount, routing), instead of creating the correct inheritance initially. #996 is not needed to solve this particular problem

**Comment by JartanFTW** [comment]
@MaximDementyev This workaround avoids the problem by keeping root_path off the app that hosts admin, but it doesn't help when root_path is set by infrastructure you don't control (reverse proxies, load balancers, Kubernetes ingress). In those cases the ASGI server injects root_path into the scope and there's no app restructuring that can prevent it. That's the scenario #996 addresses.

**Comment by MaximDementyev** [comment]
@JartanFTW I agree with yo but if the root_path is set by an infrastructure that you do not control (reverse proxies, load balancers, access to Kubernetes) How will the admin panel work in this case? Will the path to it bypass the proxy server/load balancer, and if we don't have control, how can we bypass it? Is there really such a need when different routes to the same service are required?

**Comment by JartanFTW** [comment]
@MaximDementyev A common case: your app serves both API endpoints and an admin panel, deployed behind a proxy that mounts it at /myservice. All traffic hits /myservice/... — both /myservice/api/... and /myservice/admin/.... The proxy strips /myservice, forwards the request, and sets root_path="/myservice". sqladmin needs to include that prefix in the URLs it generates (links, assets, AJAX), otherwise the browser requests /admin/statics/... instead of /myservice/admin/statics/... and gets a 404 from the proxy.

**Comment by mmzeynalli** [comment]
@MaximDementyev Firstly, I think the library should be flexible enough to accommodate both needs. It would be problematic if it was hardly maintainable code, but in our PR case, it is a simple solution. So, I do not see why we should not enable for users to have it both ways.

Plus, having a workaround is not always nice.

**Comment by MaximDementyev** [comment]
And to solve the ajax problem, wouldn't it be right to just fix the ajax url?
https://github.com/smithyhq/sqladmin/blob/5ec2360f9d17e40414f4bcda24b2dac0fcda8f9c/sqladmin/application.py#L242
Something like that 
`view.ajax_lookup_url = urljoin(self.app.root_path + self.base_url + "/", f"{view.identity}/ajax/lookup")`
The FastAPI behind Nginx seems to be working for me.

**Comment by mmzeynalli** [comment]
Not actual anymore, fixed in some release.

**Comment by smac89** [comment]
In case anyone is stuck on this, this is how I was able to come up with a solution for mounting `sqladmin` (as a starlette app) inside `Litestar` (with the help of claude):

```py
admin_app: Starlette = admin.create()

@asgi(path="/admin", is_static=True, is_mount=True, copy_scope=True)
async def admin_route(scope: "Scope", receive: "Receive", send: "Send") -> None:
    scope["root_path"] = "/admin"
    scope["path"] = scope["root_path"] + scope["path"]
    await admin_app(scope, receive, send)
```

Initially, I had just `scope["root_path"] = "/admin"` and this worked for every path except for static files. After a whole day of frustration trying various hacks I found all over the internet, I handed it over to Claude and it came up with the above solution.

When I asked Claude to explain why prefixing the `scope['path']` with `scope['root_path']` works, it gave the following explanation:

### tl;dr
> Plain routes survive Litestar's path-stripping because they match against `get_route_path` and don't propagate path info further. `Mount` + `StaticFiles` is a two-layer routing system where the inner layer relies on the prefix arithmetic done by the outer layer — that arithmetic is only correct when `scope["path"]` is the full path. Setting `root_path` alone tells admin's URL generator the right prefix to _output_, but doesn't repair the _input_ path that the nested Mount routing needs to subtract from.

<details><summary>Claude's response</summary>

Good question — it comes down to how Starlette's nested-mount routing computes paths versus how plain routes do. The two use different mechanisms, and only one of them depends on `scope["path"]` and `scope["root_path"]` being consistent.

## The core invariant
Starlette's routing follows one rule: `scope["path"]` **is always the full request path; `scope["root_path"]` is the prefix consumed so far; and the "current route path" is computed on demand via** `get_route_path(scope)` ≈ `scope["path"][len(scope["root_path"]):]` (see [starlette/_utils.py:90](vscode-webview://0do2is47p4je2n7fr8jgug30mp0j0b4flfpme52eqiep3ge8epiq/server/.venv/lib/python3.13/site-packages/starlette/_utils.py#L90)).
Litestar's `is_mount=True` violates that invariant: it strips `/admin` from `scope["path"]` _before_ handing the scope off ([litestar/_asgi/asgi_router.py:84-90](vscode-webview://0do2is47p4je2n7fr8jgug30mp0j0b4flfpme52eqiep3ge8epiq/server/.venv/lib/python3.13/site-packages/litestar/_asgi/asgi_router.py#L84-L90)). Setting `scope["root_path"]="/admin"` on its own doesn't restore consistency — `scope["path"]` no longer starts with `scope["root_path"]`.

## Why plain admin routes still work
A Route matches by running its regex against `get_route_path(scope)`. `get_route_path` has a safety net for this exact mismatch:

```py
def get_route_path(scope):
    path = scope["path"]
    root_path = scope.get("root_path", "")
    if not root_path:
        return path
    if not path.startswith(root_path):
        return path        # ← falls through silently
    ...
```

So when `path="/users/list"` and `root_path="/admin"`, it returns `/users/list`. The regex on `Route("/{identity}/list")` matches that fine. The endpoint runs. **The mismatch never gets noticed** because plain routes don't propagate any path info to a nested app — they just call their endpoint function.

`request.url_for("admin:statics", path=...)` also works because URL generation only consults `scope["root_path"]` (and `app_root_path`) to know the prefix to prepend — it doesn't have to _consume_ the prefix from `scope["path"]`.

## Why the `/statics` mount + `StaticFiles` breaks

`Mount` is different. Look at [starlette/routing.py:387-417](vscode-webview://0do2is47p4je2n7fr8jgug30mp0j0b4flfpme52eqiep3ge8epiq/server/.venv/lib/python3.13/site-packages/starlette/routing.py#L387-L417):

```py
def matches(self, scope):
    root_path = scope.get("root_path", "")
    route_path = get_route_path(scope)
    match = self.path_regex.match(route_path)
    if match:
        remaining_path = "/" + matched_params.pop("path")
        matched_path = route_path[: -len(remaining_path)]
        child_scope = {
            ...
            "root_path": root_path + matched_path,   # ← additive
            ...
        }
```

`Mount` **appends** its matched prefix to the child's `root_path`. The child app then expects `scope["path"] - scope["root_path"]` to equal "the part inside the mount."

Trace the failing request without the `scope["path"]` fix:

Layer | scope["path"] | scope["root_path"] | get_route_path()
-- | -- | -- | --
Litestar strips `/admin` | `/statics/css/main.css` | `""` | `/statics/css/main.css`
`admin_route` sets root_path | `/statics/css/main.css` | `/admin` | `/statics/css/main.css` (fallback)
Admin's outer Mount("/") matches → `root_path` += "" | `/statics/css/main.css` | `/admin` | `/statics/css/main.css`
Admin's `Mount("/statics")` matches → root_path += `/statics` | `/statics/css/main.css` | `/admin/statics` | `/statics/css/main.css` (still fallback!)
`StaticFiles.get_path()` calls `get_route_path` | `/statics/css/main.css` | `/admin/statics` | `/statics/css/main.css`

## With the fix

Prepending `root_path` to `scope["path"]` (i.e. `scope["path"] = "/admin" + "/statics/css/main.css"`) restores the invariant:

Layer | scope["path"] | scope["root_path"] | get_route_path()
-- | -- | -- | --
After fix | `/admin/statics/css/main.css` | `/admin` | `/statics/css/main.css`
Admin's outer `Mount("/")` → root_path += `""` | `/admin/statics/css/main.css` | `/admin` | `/statics/css/main.css`
Admin's `Mount("/statics")` → root_path += `/statics` | `/admin/statics/css/main.css` | `/admin/statics` | `/css/main.css` ✓
`StaticFiles` looks up `css/main.css` | — | — | served

Every `get_route_path` call now correctly subtracts the consumed prefix, so each layer sees only what's "inside" it.
</details>
